### PR TITLE
[libc] Add memcmp / bcmp fuzzers

### PR DESCRIPTION
**Pull request by gchatelet**

### CODE REVIEW

File: libc/fuzzing/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=fuzzer")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=fuzzer,address")
```
Comment (gchatelet) [inline libc/fuzzing/CMakeLists.txt:1]: https://llvm.org/docs/LibFuzzer.html#fuzzer-usage :

_In most cases you may want to combine libFuzzer with [AddressSanitizer](https://clang.llvm.org/docs/AddressSanitizer.html) (ASAN), [UndefinedBehaviorSanitizer](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) (UBSAN), or both._

Comment (legrosbuffle) [reply libc/fuzzing/CMakeLists.txt:1]: Let's do that in a separate patch

File: libc/fuzzing/string/bcmp_fuzz.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  return 0;
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  const auto normalize = [](int value) -> int {
+    if (value == 0)
+      return 0;
+    return 1;
+  };
+  const auto count = size / 2;
```
Comment (legrosbuffle) [inline libc/fuzzing/string/bcmp_fuzz.cpp:34]: `// We ignore the last byte is size is odd.`

File: libc/fuzzing/string/bcmp_fuzz.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    return 0;
+  const auto print = [](const char *msg, const char *buffer, size_t size) {
+    printf("%s\"", msg);
+    for (size_t i = 0; i < size; ++i)
+      printf("\\x%02x", (uint8_t)buffer[i]);
+    printf("\"\n");
+  };
+  print("a     : ", a, count);
+  print("b     : ", b, count);
+  printf("count : %zu\n", count);
```
Comment (legrosbuffle) [inline libc/fuzzing/string/bcmp_fuzz.cpp:49]: Nit: print count first ?

File: libc/fuzzing/string/memcmp_fuzz.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  const auto sign = [](int value) -> int {
+    if (value < 0)
+      return -1;
+    if (value > 0)
+      return 1;
+    return 0;
+  };
+  const auto count = size / 2;
```
Comment (legrosbuffle) [inline libc/fuzzing/string/memcmp_fuzz.cpp:39]: ditto x2

File: libc/fuzzing/string/memcmp_fuzz.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  const auto print = [](const char *msg, const char *buffer, size_t size) {
+    printf("%s\"", msg);
+    for (size_t i = 0; i < size; ++i)
+      printf("\\x%02x", (uint8_t)buffer[i]);
+    printf("\"\n");
+  };
+  print("a     : ", a, count);
+  print("b     : ", b, count);
+  printf("count : %zu\n", count);
+  printf("result: %d\n", reference);
```
Comment (legrosbuffle) [inline libc/fuzzing/string/memcmp_fuzz.cpp:55]: You probably want to print the actual result too, given that there are 3 semantically distinct values.

**Review by legrosbuffle — APPROVED** [review]

**Comment by gchatelet** [comment]
In preparation for https://github.com/llvm/llvm-project/pull/77081

File: utils/bazel/crash-f7dbdb2b330aad91f520099159e736e91bb9ddbf
Comment (rupprecht) [inline utils/bazel/crash-f7dbdb2b330aad91f520099159e736e91bb9ddbf:1]: Just noticed this file and couldn't quite figure out what its use is. It looks like maybe it's an artifact of running the fuzz test and forgetting to exclude it from the commit. Is it needed for anything, or should we remove it?

Comment (michaelrj-google) [reply utils/bazel/crash-f7dbdb2b330aad91f520099159e736e91bb9ddbf:1]: yes it's an artifact from running the fuzzer, and yes it's safe to remove.

**Pull request merged**